disorder-corpus/src/Disorder/Corpus.hs

```diff
@@ -1,19 +1,105 @@
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NoImplicitPrelude #-}
 {-# LANGUAGE OverloadedStrings #-}
 module Disorder.Corpus (
-    cooking
+    Cooking(..)
+  , unCooking
+  , cooking
+
+  , Muppet(..)
+  , unMuppet
   , muppets
+
+  , Southpark(..)
+  , unSouthpark
   , southpark
+
+  , Simpson(..)
+  , unSimpson
   , simpsons
+
+  , Virus(..)
+  , unVirus
   , viruses
+
+  , Colour(..)
+  , unColour
   , colours
+
+  , Weather(..)
+  , unWeather
   , weather
+
+  , Water(..)
+  , unWater
   , waters
+
+  , Boat(..)
+  , unBoat
   , boats
   ) where
 
-import           Data.Text
+import           Data.Data (Data)
+import           Data.Eq (Eq)
+import           Data.Functor (fmap)
+import qualified Data.List as List
+import           Data.Monoid ((<>))
+import           Data.Ord (Ord)
+import           Data.String (IsString(..))
+import           Data.Typeable (Typeable)
+
+import           GHC.Generics (Generic)
+
+import           Test.QuickCheck (Arbitrary(..), Gen, oneof, elements)
+
+import           Text.Show (Show)
+import           Text.Read (Read)
+
+
+-- | Generate something in the corpus or something completely bonkers.
+genCorpus :: [a] -> (a -> b) -> Gen a -> Gen b
+genCorpus corpus f gen =
+  oneof [
+      fmap f (elements corpus)
+    , fmap f gen
+    ]
 
-cooking :: [Text]
+-- | Shrinks 'b', preferring to return things from the corpus. If 'b' is
+--   already from the corpus then shrinks to nothing.
+shrinkCorpus :: Eq a => [a] -> (a -> b) -> (b -> a) -> (a -> [a]) -> b -> [b]
+shrinkCorpus corpus f g shrinkA b =
+  let
+    a = g b
+  in
+    if List.elem a corpus then
+      []
+    else
+      -- Items at the start of the shrink list are tried first by QuickCheck,
+      -- so put the corpus items first and hopefully we'll get a nicer
+      -- counterexample.
+      fmap f (corpus <> shrinkA a)
+
+--
+-- The newtypes below have the unXXX function defined separately so that
+-- the derived Show instance produces output which is easier to read.
+--
+
+newtype Cooking a =
+  Cooking a
+  deriving (Eq, Ord, Read, Show, Generic, Data, Typeable)
+
+unCooking :: Cooking a -> a
+unCooking (Cooking x) =
+  x
```

Comment:
Moved the unXXX functions out of the newtype so that the Show output is more readable